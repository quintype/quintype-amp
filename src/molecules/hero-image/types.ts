import { Story } from "../../types/story";
import { Config } from "../../types/config";

export interface HeroImageBaseTypes {
  story: Story;
  config: Config;
}
