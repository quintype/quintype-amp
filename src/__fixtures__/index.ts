import { config, publisherConfig, ampConfig, configOpts } from "./config.fixture";
import { textStory } from "./text-story.fixture";
import { storyWithManyJsEmbeds } from "./story-with-jsembeds.fixture";
import { relatedStories } from "./related-stories";
import { seo } from "./seo.fixture";
import { allElementsStory } from "./all-element-story.fixture";
import { liveBlogStory } from "./live-blog-story.fixture";

export {
  allElementsStory,
  config,
  publisherConfig,
  ampConfig,
  textStory,
  storyWithManyJsEmbeds,
  relatedStories,
  configOpts,
  seo,
  liveBlogStory
};
