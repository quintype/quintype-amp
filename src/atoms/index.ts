import Carousel from "./carousel";
import { Image } from "./image";
import Layout from "./layout";
import Facebook from "./facebook";
import Twitter from "./twitter";
import { Footer } from "./footer";
import { Section } from "./section";
import { Spacer } from "./spacer";
import { Author } from "./author";
import { DateTime } from "./date-time";
import { StoryElement } from "./story-elements/story-element";
import { StoryElements } from "./story-elements";

export {
  Carousel,
  Image,
  Layout,
  Facebook,
  Twitter,
  Footer,
  Section,
  Spacer,
  Author,
  DateTime,
  StoryElement,
  StoryElements
};
