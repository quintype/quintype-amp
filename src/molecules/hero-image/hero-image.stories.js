import * as React from "react";
import { storiesOf } from "@storybook/react";
import { HeroImage } from "./hero-image"; 
import { textStory } from "../../__fixtures__/story.fixture";
import { config } from "../../__fixtures__/config.fixture";
import Layout from "../../atoms/layout"

storiesOf("HeroImage", module)
  .addDecorator((story) => <Layout story={textStory} config={config}>{story()}</Layout>)
  .add("default", () => {
    return <HeroImage />
  })

