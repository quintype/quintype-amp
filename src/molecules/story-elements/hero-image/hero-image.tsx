import React, { useContext } from "react";
import Image from "../../../atoms/image";
import { StoryContext } from "../../../context/storyContext";
import { Story } from "../../../types/story";
import { HeroImageTypes, HeroImageBaseTypes } from "./types";
import styled from "styled-components";
import { media } from "../../../utils/media";

const StyledFigcaption = styled.figcaption`
  text-align: left;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  padding: 8px;
  max-height: 90px;
  color: ${(props) => props.theme.color.white};
  line-height: ${(props) => props.theme.font.lineHeight.level1};
  background: ${(props) => `${props.theme.color.black}60`};
  font-size: ${(props) => props.theme.font.size.xxs};
  z-index: ${(props) => props.theme.zIndex.z100};

  ${media.laptop`
		max-height: 130px;
		overflow-y: scroll;
	`}
`;

export const HeroImage = (props: HeroImageTypes) => {
  const storyFromContext = useContext(StoryContext) as Story;
  return <HeroImageBase story={storyFromContext} {...props} />;
};

export const HeroImageBase = ({ story, attribution, slug, metadata, caption }: HeroImageBaseTypes) => {
  let overRideStory = false;
  let figcaptionText: string | undefined | boolean;
  const imageProps: any = {
    aspectRatio: [1200, 750]
  };
  if (attribution || slug || metadata || caption) overRideStory = true;
  if (overRideStory) {
    imageProps.alt = caption;
    imageProps.metadata = metadata;
    imageProps.slug = slug;
    imageProps.attribution = attribution;
    figcaptionText = getFigcaptionText(caption, attribution);
  } else {
    const {
      "hero-image-attribution": ATTRIBUTION,
      "hero-image-s3-key": SLUG,
      "hero-image-metadata": METADATA,
      "hero-image-caption": CAPTION
    } = story;
    imageProps.alt = CAPTION;
    imageProps.metadata = METADATA;
    imageProps.slug = SLUG;
    imageProps.attribution = ATTRIBUTION;
    figcaptionText = getFigcaptionText(CAPTION, ATTRIBUTION);
  }
  return (
    <div>
      <Image {...imageProps}>{figcaptionText && <StyledFigcaption>{figcaptionText}</StyledFigcaption>}</Image>
    </div>
  );
};

export function getFigcaptionText(caption, attribution) {
  if (caption && attribution) return `${caption} | ${attribution}`;
  else if (caption || attribution) return `${caption || attribution}`;
  else return false;
}
