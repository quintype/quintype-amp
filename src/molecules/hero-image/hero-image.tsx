import React from "react";
import { HeroImageBaseTypes } from "./types";
import { HeroImageMetadata } from "../../types/story";
import styled from "styled-components";
import { withStoryAndConfig } from "../../context";
import { Image } from "../../atoms";
import get from "lodash.get";

const StyledText = styled.span`
  margin-right: 8px;
  color: ${(props) => props.theme.color.mono6};
  line-height: ${(props) => props.theme.font.lineHeight.level1};
  font-size: ${(props) => props.theme.font.size.xxs};
  word-wrap: break-word;
`;

const StyledAttributionText = styled.span`
   color: ${(props) => props.theme.color.mono4};
   font-size: ${(props) => props.theme.font.size.xxs};
`;
const StyledDiv = styled.div`
  background-color: ${(props) => props.theme.color.mono2};
`;
const StyledFigcaption = styled.div`
  text-align: left;
  padding-top: 8px;
`;
export const HeroImageBase = ({ story }: HeroImageBaseTypes) => {
  const metadata: HeroImageMetadata = get(story, "hero-image-metadata", null);
  const slug: string | null = get(story, "hero-image-s3-key", null);
  if (!slug || !metadata) return null;

  const attribution: string | null = get(story, "hero-image-attribution", null);
  const caption: string | null = get(story, "hero-image-caption", null);

  return (
   <>
      <StyledDiv>
        <Image data-hero="true" metadata={metadata} slug={slug} alt={caption || attribution || ""}>
        </Image>
      </StyledDiv>
      <StyledFigcaption>
        {caption && <StyledText dangerouslySetInnerHTML={{ __html: `${caption}` + '&nbsp;' }} />}
        {attribution && <StyledAttributionText dangerouslySetInnerHTML={{ __html: `${attribution}` }} />}
      </StyledFigcaption>
    </>
  );
};

/**
 * HeroImage Component is a wrapper around the <amp-img> amp component. It even serves figure caption text which picks up "hero-image-attribution" and/or "hero-image-caption" depending which one is available.
 *
 * @category Molecules
 * @component
 */

export const HeroImage = withStoryAndConfig(HeroImageBase);

export function getFigcaptionText(caption, attribution) {
  if (caption && attribution) return `${caption} | ${attribution}`;
  else if (caption || attribution) return `${caption || attribution}`;
  else return false;
}
