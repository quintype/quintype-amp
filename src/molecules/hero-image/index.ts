import { HeroImage } from "./hero-image";
export { HeroImage };
export default HeroImage;
