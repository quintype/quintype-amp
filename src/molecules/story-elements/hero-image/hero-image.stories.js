import * as React from "react";
import { storiesOf } from "@storybook/react";
import { HeroImage } from "./hero-image";
import { textStory } from "../../../__fixtures__/story.fixture";
import { config } from "../../../__fixtures__/config.fixture";
import Layout from "../../../atoms/layout";

const customMetadata = {
  width: 2560,
  height: 960,
  "mime-type": "image/jpeg",
  "file-size": 863219,
  "file-name": "petition filed in Delhi HC.jpg",
  "focus-point": [1249, 409]
};
const customS3Key = "barandbench/2020-02/0ecdae82-a09e-47a9-9249-d8d1380e0de8/petition_filed_in_Delhi_HC.jpg";

storiesOf("HeroImage", module)
  .addDecorator((story) => (
    <Layout story={textStory} config={config}>
      {story()}
    </Layout>
  ))
  .add("default", () => {
    return <HeroImage />;
  })
  .add("HeroImage with extremely long caption", () => {
    const junk =
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam tincidunt ut neque vel luctus. In dictum imperdiet tellus, vitae vestibulum elit dictum porta. Vivamus vel velit eu arcu tristique finibus et sit amet nunc. Donec condimentum lobortis nisl a vulputate. Pellentesque interdum scelerisque justo, a commodo lorem laoreet id. Curabitur molestie dictum velit, sed vehicula lorem ornare quis. Nam ac dui a velit molestie ullamcorper. Curabitur pharetra justo at tellus venenatis tincidunt. Integer bibendum ante vel vulputate tincidunt. Aliquam quis tortor id justo dignissim lacinia. Sed ligula nulla, convallis ornare tincidunt vitae, venenatis sed turpis. Nam imperdiet finibus felis vel euismod. Fusce tortor sapien, molestie id turpis non, lobortis venenatis felis. Duis facilisis dictum turpis, ac pharetra nunc imperdiet sit amet. Nullam augue orci, vestibulum aliquam augue at, pulvinar dictum justo. Vivamus nec lobortis nisi. Integer fermentum aliquam massa eu volutpat. Vivamus sit amet leo in metus euismod mollis et sed magna. Suspendisse luctus pretium libero, vitae vehicula lacus vehicula a. Vivamus elementum magna at interdum porta. Curabitur condimentum dui ut sapien luctus, id mattis elit sodales. Etiam bibendum dolor quis sapien pellentesque, at interdum mi laoreet. Nulla tellus augue, convallis non lacinia sed, placerat nec magna. Quisque blandit velit ac leo rutrum, at euismod sapien fringilla. Phasellus hendrerit tincidunt varius. In vehicula blandit pharetra. In vel pharetra diam, nec mattis purus. Maecenas mollis eu erat non aliquet. Phasellus a ligula ac quam faucibus congue sed ut enim. Donec vehicula leo vitae odio semper consectetur non lacinia velit. Pellentesque feugiat leo ut est ornare accumsan. Etiam at condimentum nunc. Vivamus sit amet leo in metus euismod mollis et sed magna. Suspendisse luctus pretium libero, vitae vehicula lacus vehicula a. Vivamus elementum magna at interdum porta. Curabitur condimentum dui ut sapien luctus, id mattis elit sodales. Etiam bibendum dolor quis sapien pellentesque, at interdum mi laoreet. Nulla tellus augue, convallis non lacinia sed, placerat nec magna. Quisque blandit velit ac leo rutrum, at euismod sapien fringilla. Phasellus hendrerit tincidunt varius. In vehicula blandit pharetra. In vel pharetra diam, nec mattis purus. Maecenas mollis eu erat non aliquet. Phasellus a ligula ac quam faucibus congue sed ut enim. Donec vehicula leo vitae odio semper consectetur non lacinia velit. Pellentesque feugiat leo ut est ornare accumsan. Etiam at condimentum nunc.";
    return <HeroImage caption={junk} attribution={null} slug={customS3Key} metadata={customMetadata} />;
  })
  .add("HeroImage with custom attributes - with no attribution", () => {
    return (
      <HeroImage
        caption=" Congress, AAP and AIMIM leaders "
        attribution={null}
        slug={customS3Key}
        metadata={customMetadata}
      />
    );
  })
  .add("HeroImage with custom attributes - with no caption", () => {
    return <HeroImage caption={null} attribution="custom attribution" slug={customS3Key} metadata={customMetadata} />;
  })
  .add("HeroImage with custom attributes", () => {
    return (
      <HeroImage
        caption=" Congress, AAP and AIMIM leaders "
        attribution="custom attribution"
        slug={customS3Key}
        metadata={customMetadata}
      />
    );
  })
  .add("HeroImage with custom attributes - with no caption and no attribution", () => {
    return <HeroImage caption={null} attribution={null} slug={customS3Key} metadata={customMetadata} />;
  });
