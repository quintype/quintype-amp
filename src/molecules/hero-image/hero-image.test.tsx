import React from "react";
import { HeroImage } from "./hero-image";
import { mount } from "enzyme";
import { getFigcaptionText } from "./hero-image";
import { textStory, config } from "../../__fixtures__";
import { Layout } from "../../atoms";

describe("HeroImage", () => {
  it("Should add data-hero attribute on amp-img so that amp optimizer will preload it", () => {
    const wrapper = mount(
      <Layout story={textStory} config={config}>
        <HeroImage />
      </Layout>
    );
    expect(wrapper.find("amp-img").prop("data-hero")).toBe("true");
  });
});

test("getFigcaptionText should return caption and attribution", () => {
  const text = getFigcaptionText("img caption", "img attribution");
  expect(text).toBe("img caption | img attribution");
});
test("getFigcaptionText should return attribution", () => {
  const text = getFigcaptionText(null, "img attribution");
  expect(text).toBe("img attribution");
});
test("getFigcaptionText should return caption", () => {
  const text = getFigcaptionText("img caption", null);
  expect(text).toBe("img caption");
});
test("getFigcaptionText should return false", () => {
  const text = getFigcaptionText(null, null);
  expect(text).toBe(false);
});
test("getFigcaptionText should return caption and attribution without HTML Tags", () => {
  const text = getFigcaptionText("img caption", "img attribution");
  expect(text).toMatchSnapshot();
});
