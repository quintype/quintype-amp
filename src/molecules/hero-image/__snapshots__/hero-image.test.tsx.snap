// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroImage should render 1`] = `
<Component
  config={Object {}}
  story={Object {}}
/>
`;

exports[`HeroImage should render caption without HTML Tags 1`] = `
<Component
  attribution="<span>custom attribution</span>"
  caption="<p>Congress, AAP and AIMIM leaders</p>"
  config={Object {}}
  story={Object {}}
/>
`;

exports[`getFigcaptionText should return caption and attribution without HTML Tags 1`] = `"img caption | img attribution"`;
